### PR TITLE
Enhance `nats_kv` processor

### PR DESCRIPTION
**Pull request by mihaitodor**
This is a slightly-breaking change to the nats_kv processor which is still beta. It alters the behaviour of the `keys` operation which now returns an array of keys instead of a batch of messages, as discussed with @codegangsta.

I'm a bit concerned that the history operation still returns a batch of messages, which might confuse people. Happy to modify that one as well to return an array of objects if you'd like. In the current implementation, some of the fields of each history item are emitted as message metadata, which is a bit odd anyway, since they're not identical across the batch...

PS: I also took the liberty to clean up the code a tiny bit.

### CODE REVIEW

File: internal/impl/nats/processor_kv.go
Comment (mihaitodor) [inline internal/impl/nats/processor_kv.go:336]: I'm a bit concerned about this approach, because it seems to block indefinitely in some cases. For example, try setting `keys_filter: '"001.>"'`. Not sure if this is by design, or a bug, but I think we might need to consider adding a `timeout` field that lets us error out if a timer expires while we wait for updates. WDYT?

File: internal/impl/nats/processor_kv.go
Comment (mihaitodor) [inline internal/impl/nats/processor_kv.go:304]: I just noticed that `kv.History()` supports wildcards in the `key`. Not sure if this is by design @codegangsta, but I think it would be better to adjust the docs and pass in `keys_filter` instead of `key` for this operation as well. Also, see the comment about changing this one to return an array of objects instead of a batch of messages.

File: internal/impl/nats/processor_kv.go
Comment (mihaitodor) [inline internal/impl/nats/processor_kv.go:321]: Does it make sense to return (or log) an error here? The Nats client library ignores it completely: https://github.com/nats-io/nats.go/blob/46be7408f6630569929044fcad459ba0844a296d/kv.go#L870

**Comment by codegangsta** [comment]
Taking a peek at this right now

File: internal/impl/nats/processor_kv.go
Comment (codegangsta) [reply internal/impl/nats/processor_kv.go:343]: Interesting you are getting blocking on that value, I wonder if something is breaking when you send NATS a subject it doesn't like.

All that said, it's entirely possible for the watcher to never end, depending on how fast keys are being inserted. More of an edge case, but one concession we could make is to either have a configurable timeout, or stop the watcher when a specific time is reached, or a combination of the two.

I don't particularly like the idea that a simple timeout would result in receiving a partial set of keys. so I agree an error would be best in the simple timeout case.

That said, I kinda do like the idea of just returning the keys up to time that we requested it, would result in potentially more stale data for more timeliness on the retrieval side.

This is definitely more of an edge case where we have a really hot kv at the moment the watcher is created. I'm cool with any of the solutions.

File: internal/impl/nats/processor_kv.go
Comment (codegangsta) [reply internal/impl/nats/processor_kv.go:328]: Most of the errors have to do with race conditions, having to do with:

1. The connection being closed or draining
2. The underlying subscription already being unsubscribed or
3. The underlying consumer already being deleted

All possible but not particularly actionable for this use case, and generally exceptional. **Maybe** a DEBUG log could be useful here someday but I can't think of a reason why it would be at the moment.

File: website/docs/components/processors/nats_kv.md
Comment (codegangsta) [inline website/docs/components/processors/nats_kv.md:229]: I'm going back and forth with the idea of introducing this vs reusing the `key` field. My hangup is that at the end of the day they are serving the same purpose for any `watch` related options, so `History`, `Keys` and `Watch` in the input. I wonder if the `key` field should be used for all of it. WDYT

File: internal/impl/nats/processor_kv.go
Comment (codegangsta) [reply internal/impl/nats/processor_kv.go:304]: We should absolutely make it consistent, I added some questions about the use of `keys_filter` in another comment to discuss.

File: internal/impl/nats/processor_kv.go
Comment (mihaitodor) [reply internal/impl/nats/processor_kv.go:343]: I think I'll go with the more conservative route and just return an error if the timeout lapses. People should take that as a sign that they either need to bump up the timeout or use a wildcard which selects fewer keys. We could enable partial results later on with a separate flag if anyone needs that functionality.

I'll slap together some quick code which shows how `'"001.>"'` blocks and I'll open an issue here https://github.com/nats-io/nats.go.

**Review by codegangsta — CHANGES_REQUESTED** [review]
Added some questions and thoughts.

One overall thought is that while we are here, we should at least upgrade to the new version of the jetstream client, which is now in it's own package: `github.com/nats-io/nats.go/jetstream`.

For KV there shouldn't be much in the way of changes, I think we can circle back on the other jetstream components in another PR, as the more significant changes are in how consumers work.

I can ofc help with the implementation, just lmk

File: internal/impl/nats/processor_kv.go
Comment (mihaitodor) [reply internal/impl/nats/processor_kv.go]: Sounds good, I added a debug log just in case someone needs to troubleshoot it.

File: internal/impl/nats/processor_kv.go
Comment (mihaitodor) [reply internal/impl/nats/processor_kv.go:345]: Issue opened here: https://github.com/nats-io/nats.go/issues/1608

**Comment by mihaitodor** [comment]
@codegangsta I updated the `nats_kv` components code to use the new `github.com/nats-io/nats.go/jetstream` in 5c4f02f68. It's nice to see that the API calls are context-aware now. Wasn't sure if it's ready for prime-time, but it seems to work well. I'll leave the upgrade of the `nats_jetstream` components for a separate PR.

I have a few more comments to address. Stay tuned :)

File: website/docs/components/processors/nats_kv.md
Comment (mihaitodor) [reply website/docs/components/processors/nats_kv.md:229]: Yeah, I'm leaning towards that as well. My initial thought was that it might be confusing for users if they set a wildcard for operations which don't support them, but they should get an error in that case anyway. I'll just document that the `history` and `keys` operations support wildcards. Should've done this from the start.

**Comment by mihaitodor** [comment]
OK, I went ahead and also adjusted the `history` operation to return a list of objects instead of a batch of messages and documented the fields that are returned. See 1a8d57b. Guess I can DRY up the code a bit more if need be, but I'm curious if you think this breaking change makes sense.

File: internal/impl/nats/processor_kv.go
Comment (mihaitodor) [reply internal/impl/nats/processor_kv.go:344]: And I also added a `timeout` field in 83c662e.

Comment (codegangsta) [reply internal/impl/nats/processor_kv.go:344]: I think that's a good solution for most cases!

File: website/docs/components/processors/nats_kv.md
Comment (codegangsta) [reply website/docs/components/processors/nats_kv.md]: Yeah. Looking at the code I'm leaning more toward that! Nice call

**Review by codegangsta — APPROVED** [review]
Looking good!

Wish there was a release when this is merged since I had to do some ungodly acts in my next video to get keys. Thankfully I won't be exposing any breaking changes there. Thanks so much for doing this. Hit me up on discord if you are interested in doing the `nats_jetstream` component updates since I do infact have a lot of opinions there, and may be working on some this week if you want to move on to other things.

Thanks again for updating to the newer jetstream api

**Comment by mihaitodor** [comment]
Awesome, thanks @codegangsta! I'll defer to @Jeffail for the rest of the review. Hope he's OK with the small breaking changes :)

> Hit me up on discord if you are interested in doing the `nats_jetstream` component updates since I do infact have a lot of opinions there, and may be working on some this week if you want to move on to other things.

Sounds good! I might take a stab at it this week.

**Review by Jeffail — APPROVED** [review]
LGTM, thanks @mihaitodor!

**Pull request merged**
